vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/ConvertDocExpressionToFile.java

```diff
@@ -0,0 +1,171 @@
+        // TODO Auto-generated method stub
+        return false;
+    }
+
+    @Override
+    public boolean rewritePost(Mutable<ILogicalOperator> opRef, IOptimizationContext context)
+            throws AlgebricksException {
+        boolean modified = false;
+        List<Mutable<ILogicalExpression>> expressions = OperatorToolbox.getExpressions(opRef);
+        for (Mutable<ILogicalExpression> expression : expressions) {
+            Mutable<ILogicalExpression> docExpression = ExpressionToolbox.findFirstFunctionExpression(expression,
+                    BuiltinFunctions.FN_DOC_1.getFunctionIdentifier());
+            if (docExpression != null) {
+                AbstractFunctionCallExpression absFnCall = (AbstractFunctionCallExpression) docExpression.getValue();
+                if (docExpression != null && ifDocExpressionFound(opRef, absFnCall.getArguments().get(0), context)) {
+                    modified = true;
+                }
+            }
+        }
+        return modified;
+    }
+
+    protected boolean ifDocExpressionFound(Mutable<ILogicalOperator> opRef, Mutable<ILogicalExpression> funcExpression,
+            IOptimizationContext context) {
+        VXQueryConstantValue constantValue = null;
+        ConstantExpression constantExpression = null;
+        ILogicalExpression logicalExpression = (ILogicalExpression) funcExpression.getValue();
+
+        if (logicalExpression.getExpressionTag() == LogicalExpressionTag.CONSTANT) {
+
+            constantExpression = (ConstantExpression) logicalExpression;
+            constantValue = (VXQueryConstantValue) constantExpression.getValue();
+        } else if (logicalExpression.getExpressionTag() == LogicalExpressionTag.VARIABLE) {
+
+            VariableReferenceExpression varLogicalExpression = (VariableReferenceExpression) logicalExpression;
+            Mutable<ILogicalOperator> lop = OperatorToolbox.findProducerOf(opRef,
+                    varLogicalExpression.getVariableReference());
+            ILogicalExpression variableLogicalExpression = (ILogicalExpression) OperatorToolbox.getExpressionOf(lop,
+                    varLogicalExpression.getVariableReference()).getValue();
+            if (variableLogicalExpression.getExpressionTag() != LogicalExpressionTag.CONSTANT) {
+                return false;
+            }
+            constantExpression = (ConstantExpression) variableLogicalExpression;
+            constantValue = (VXQueryConstantValue) constantExpression.getValue();
+        } else {
+            return false;
+        }
+
+        if (constantValue.getType() != SequenceType.create(BuiltinTypeRegistry.XS_STRING, Quantifier.QUANT_ONE)) {
+            return false;
+        }
+        tvp.set(constantValue.getValue(), 0, constantValue.getValue().length);
+        String collectionName = null;
+        tvp.getValue(stringp);
+        if (tvp.getTag() != ValueTag.XS_STRING_TAG) {
+            return false;
+        }
+        try {
```

Comment:
Since you have the stringp (string pointable), I think it would be better to use the UTF8StringPointable.toString(StringBuilder buffer) method to get the String for the doc argument. It may remove the catch and use fewer variables.